<a href="https://colab.research.google.com/github/Mhtag/tutorials/blob/main/machine_learning/L1_and_L2_regularization.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/codebasics/py/master/ML/16_regularization/Melbourne_housing_FULL.csv')
df.head()

Suburb             Address  ...             Regionname Propertycount
0  Abbotsford       68 Studley St  ...  Northern Metropolitan        4019.0
1  Abbotsford        85 Turner St  ...  Northern Metropolitan        4019.0
2  Abbotsford     25 Bloomburg St  ...  Northern Metropolitan        4019.0
3  Abbotsford  18/659 Victoria St  ...  Northern Metropolitan        4019.0
4  Abbotsford        5 Charles St  ...  Northern Metropolitan        4019.0

[5 rows x 21 columns]

In [3]:
df.shape

(34857, 21)

In [4]:
df.nunique()

Suburb             351
Address          34009
Rooms               12
Type                 3
Price             2871
Method               9
SellerG            388
Date                78
Distance           215
Postcode           211
Bedroom2            15
Bathroom            11
Car                 15
Landsize          1684
BuildingArea       740
YearBuilt          160
CouncilArea         33
Lattitude        13402
Longtitude       14524
Regionname           8
Propertycount      342
dtype: int64

In [5]:

# let's use limited columns which makes more sense for serving our purpose
cols_to_use = ['Suburb', 'Rooms', 'Type', 'Method', 'SellerG', 'Regionname', 'Propertycount', 
               'Distance', 'CouncilArea', 'Bedroom2', 'Bathroom', 'Car', 'Landsize', 'BuildingArea', 'Price']
df = df[cols_to_use]
df.head()

Suburb  Rooms Type Method  ...  Car Landsize  BuildingArea      Price
0  Abbotsford      2    h     SS  ...  1.0    126.0           NaN        NaN
1  Abbotsford      2    h      S  ...  1.0    202.0           NaN  1480000.0
2  Abbotsford      2    h      S  ...  0.0    156.0          79.0  1035000.0
3  Abbotsford      3    u     VB  ...  1.0      0.0           NaN        NaN
4  Abbotsford      3    h     SP  ...  0.0    134.0         150.0  1465000.0

[5 rows x 15 columns]

In [6]:
df.isnull().sum()

Suburb               0
Rooms                0
Type                 0
Method               0
SellerG              0
Regionname           3
Propertycount        3
Distance             1
CouncilArea          3
Bedroom2          8217
Bathroom          8226
Car               8728
Landsize         11810
BuildingArea     21115
Price             7610
dtype: int64

In [8]:

# Some feature's missing values can be treated as zero (another class for NA values or absence of that feature)
# like 0 for Propertycount, Bedroom2 will refer to other class of NA values
# like 0 for Car feature will mean that there's no car parking feature with house
cols_to_fill_zero = ['Propertycount', 'Distance', 'Bedroom2', 'Bathroom', 'Car']
df[cols_to_fill_zero] = df[cols_to_fill_zero].fillna(0)
df.isnull().sum()

Suburb               0
Rooms                0
Type                 0
Method               0
SellerG              0
Regionname           3
Propertycount        0
Distance             0
CouncilArea          3
Bedroom2             0
Bathroom             0
Car                  0
Landsize         11810
BuildingArea     21115
Price             7610
dtype: int64

In [9]:
df['Landsize'] = df['Landsize'].fillna(df.Landsize.mean())
df['BuildingArea'] = df['BuildingArea'].fillna(df.BuildingArea.mean())

In [11]:
df.isnull().sum()

Suburb              0
Rooms               0
Type                0
Method              0
SellerG             0
Regionname          3
Propertycount       0
Distance            0
CouncilArea         3
Bedroom2            0
Bathroom            0
Car                 0
Landsize            0
BuildingArea        0
Price            7610
dtype: int64

In [12]:
df.dropna(inplace=True)
df.isnull().sum()

Suburb           0
Rooms            0
Type             0
Method           0
SellerG          0
Regionname       0
Propertycount    0
Distance         0
CouncilArea      0
Bedroom2         0
Bathroom         0
Car              0
Landsize         0
BuildingArea     0
Price            0
dtype: int64

In [13]:
df = pd.get_dummies(df, drop_first=True)
df.head()

Rooms  ...  CouncilArea_Yarra Ranges Shire Council
1      2  ...                                       0
2      2  ...                                       0
4      3  ...                                       0
5      3  ...                                       0
6      4  ...                                       0

[5 rows x 745 columns]

In [14]:
X =  df.drop('Price', axis=1)
y = df.Price

In [18]:
X.head()

Rooms  ...  CouncilArea_Yarra Ranges Shire Council
1      2  ...                                       0
2      2  ...                                       0
4      3  ...                                       0
5      3  ...                                       0
6      4  ...                                       0

[5 rows x 744 columns]

In [20]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y , test_size=0.3, random_state=2)

In [21]:
from sklearn.linear_model import LinearRegression
model_reg = LinearRegression()
model_reg.fit(X_train, y_train)


LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [22]:
model_reg.score(X_test, y_test)

0.13853683161569574

In [23]:
model_reg.score(X_train, y_train)

0.6827792395792723

Here training score is 68% but test score is 13.85% which is very low

Normal Regression is clearly overfitting the data, let's try other models
Using  **Lasso (L1 Regularized) Regression Model**

## Lasso L1 regression

In [29]:
from sklearn import linear_model
model_lasso = linear_model.Lasso(alpha=50, tol=0.1)
model_lasso.fit(X_train, y_train)

Lasso(alpha=50, copy_X=True, fit_intercept=True, max_iter=1000, normalize=False,
      positive=False, precompute=False, random_state=None, selection='cyclic',
      tol=0.1, warm_start=False)

In [30]:
model_lasso.score(X_test, y_test)

0.6636280170612745

In [31]:
model_lasso.score(X_train, y_train)

0.6767149418617554

## L2 RidgeRegression

In [32]:
from sklearn.linear_model import Ridge

model_ridge = Ridge(alpha=50, tol=0.1).fit(X_train, y_train)

In [33]:
model_ridge.score(X_test, y_test)

0.6670848945194958

In [34]:
model_ridge.score(X_train, y_train)

0.6622376739684328

We see that Lasso and Ridge Regularizations prove to be beneficial when our Simple Linear Regression Model overfits. These results may not be that contrast but significant in most cases.Also that L1 & L2 Regularizations are used in Neural Networks too